opt_adapt/opt.py

```diff
@@ -92,11 +92,88 @@ def _gradient_descent(it, forward_run, m, params, u, u_, dJ_, Rspace=False):
 
     # Take a step downhill
     u -= lr * dJ
-    yield {"lr": lr, "u+": u, "u-": u_, "dJ-": dJ_}
+    yield {"lr": lr, "u+": u, "u-": u_, "dJ-": dJ_, "B": None}
+
+
+def _BFGS(it, forward_run, m, params, u, u_, dJ_, B, Rspace=False):
+    """
+    A second order routine
+    """
+    J, u = forward_run(m, u, **params.model_options)
+    dJ = fd_adj.compute_gradient(J, fd_adj.Control(u))
+    yield {"J": J, "u": u.copy(deepcopy=True), "dJ": dJ.copy(deepcopy=True)}
+    
+    # if B is in R-space, B always be none
+    if B is None:
+        if Rspace:
+            if u_ is None or dJ_ is None:
+                lr = params.lr
+            else:
+                dJ_ = fd.Function(dJ).assign(dJ_)
+                u_ = fd.Function(u).assign(u_)
+                s = u.copy(deepcopy=True)
+                s -= u_
+                y = dJ.copy(deepcopy=True)
+                y -= dJ_
+                lr = s.dat.data[0]/y.dat.data[0]
+            u -= lr * dJ
+            yield {"lr": lr, "u+": u, "u-": u_, "dJ-": dJ_, "B": None}
+        else:
+            B = Matrix(u.function_space())
+            lr = 1
+            P = B.solve(dJ)
+            u -= lr * P
+            yield {"lr": lr, "u+": u, "u-": None, "dJ-": None, "B": B}
+    
+    else:
+        dJ_ = params.transfer_fn(dJ_, dJ.function_space())
+        u_ = params.transfer_fn(u_, u.function_space())
+        
+        s = params.transfer_fn(u - u_, u.function_space())
+        y = params.transfer_fn(dJ - dJ_, dJ.function_space())
```

Comment:
`u` and `u_` should already be in the right space, so you can just do
```
s = u.copy(deepcopy=True)
s -= u_
```
and similarly for `y` based on `dJ` and `dJ_`.